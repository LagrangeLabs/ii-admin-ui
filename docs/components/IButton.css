.main-row1 {
  display: flex;
  margin-bottom: 40px;
}
.main-row1>* {
  margin-right: 26px;
}
.main-hover {
  opacity: 0.8;
}
.main-row2 {
  display: flex;
}
.main-row2 >* {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  margin-right: 26px;
}
.main-row2 p {
  font-size: 13px;
}
.main-item {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 11px;
  height: 40px;
}
.minor-row {
  display: flex;
  align-items: center;
  margin-bottom: 17px;
}
.minor-item {
  display: flex;
}
.minor-item >* {
  margin-right: 20px;
}
.minor-item-hover :not(.ant-btn-link) {
  color: #0085ff;
  border-color: #0085ff;
}
.minor-item-hover> .ant-btn-link > .anticon-edit + span,  .minor-item-hover> .ant-btn-link > .anticon-upload + span{
  color: #0085ff;
}
.minor-primary, .minor-disabled {
  margin-right: 18px;
}
.minor-hover {
  margin-right: 10px;
}