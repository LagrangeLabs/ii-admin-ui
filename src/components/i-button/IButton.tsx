import React, { FC, useEffect, useMemo, useState } from 'react';
import { Button } from 'antd';
import classNames from 'classnames';
import type { ButtonProps } from 'antd/lib/button';

import './index.less';
export interface IButtonProps extends Omit<ButtonProps, 'size'> {
  /** 设置按钮的样式名 */
  className?: string;
  /** 设置按钮的大小 */
  size?:
    | 'xl'
    | 'default'
    | 'lg'
    | 'md'
    | 'sm'
    | 'xs'
    | 'xss'
    | ButtonProps['size'];
}

export const IButton: FC<IButtonProps> = (props) => {
  const { size, className, children, ...restProps } = props;

  const buttonCls = classNames('ii-button', className);

  const mapSize = new Map();
  mapSize.set('xl', [20, 40]);
  mapSize.set('default', [20, 36]);
  mapSize.set('lg', [15, 32]);
  mapSize.set('md', [14, 30]);
  mapSize.set('sm', [14, 28]);
  mapSize.set('xs', [9, 26]);
  mapSize.set('xss', [9, 24]);
  mapSize.set('large', [20, 40]);
  mapSize.set('middle', [20, 36]);
  mapSize.set('small', [9, 22]);

  const style = useMemo(
    () => ({
      height:
        `${mapSize.get(size)?.[1]}px` || `${mapSize.get('default')?.[1]}px`,
      padding:
        `0 ${mapSize.get(size)?.[0]}px` || `0 ${mapSize.get('default')?.[0]}px`,
      borderRadius: mapSize.get(size)?.[1] > 30 ? '4px' : '5px',
      ...restProps.style,
    }),
    [size],
  );

  return (
    <Button
      data-testid="test-button"
      className={buttonCls}
      {...restProps}
      style={style}
    >
      {children}
    </Button>
  );
};

export default IButton;
