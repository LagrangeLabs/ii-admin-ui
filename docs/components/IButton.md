---
title: IButton组件
---

# IButton 组件

主行动按钮：用于主行动点，一个操作区域只能有一个主按钮

Demo:

```tsx
import React from 'react';
import { IButton } from 'ii-admin-ui';
import {
  PlusOutlined,
  VerticalAlignBottomOutlined,
  EditOutlined,
  UploadOutlined,
} from '@ant-design/icons';

import './IButton.css';

export default () => {
  return (
    <>
      <div className="main-row1">
        <IButton type="primary">提交表单</IButton>
        <IButton type="primary">
          <PlusOutlined />
          添加时间
        </IButton>
        <IButton type="primary" loading>
          Loading
        </IButton>
        <IButton className="main-hover" type="primary">
          hover状态
        </IButton>
        <IButton type="primary" disabled>
          禁用状态
        </IButton>
      </div>

      <div className="main-row2">
        <div>
          <div className="main-item">
            <IButton type="primary" size="xl">
              提交表单
            </IButton>
          </div>
          <p>高度：40px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton type="primary" size="default">
              提交表单
            </IButton>
          </div>
          <p>高度：36px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton type="primary" size="lg">
              提交表单
            </IButton>
          </div>
          <p>高度：32px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton size="md">提交表单</IButton>
          </div>
          <p>高度：30px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton size="sm">提交表单</IButton>
          </div>
          <p>高度：28px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton size="xs">提交表单</IButton>
          </div>
          <p>高度：26px</p>
        </div>
        <div>
          <div className="main-item">
            <IButton size="xss">提交表单</IButton>
          </div>
          <p>高度：24px</p>
        </div>
      </div>
    </>
  );
};
```

次要按钮：以线框按钮为主，默认按钮：用于没有主次之分的一组行动点。虚线按钮：常用于添加操作。文本按钮：用于最次级的行动点。 链接按钮：用于作为外链的行动点。

Demo:

```tsx
import React from 'react';
import { IButton } from 'ii-admin-ui';
import {
  PlusOutlined,
  VerticalAlignBottomOutlined,
  EditOutlined,
  UploadOutlined,
} from '@ant-design/icons';

import './IButton.css';

export default () => {
  return (
    <>
      <div className="minor-row">
        <span className="minor-primary">常规状态</span>
        <div className="minor-item">
          <IButton>
            <VerticalAlignBottomOutlined />
            下载
          </IButton>
          <IButton type="dashed">
            <PlusOutlined />
            添加文件
          </IButton>
          <IButton>
            <UploadOutlined />
            导出
          </IButton>
          <IButton>增加时间</IButton>
          <IButton type="link">
            <EditOutlined />
            编辑
          </IButton>
          <IButton type="link">
            <UploadOutlined />
            上传
          </IButton>
        </div>
      </div>
      <div className="minor-row">
        <span className="minor-hover">hover状态</span>
        <div className="minor-item minor-item-hover">
          <IButton>
            <VerticalAlignBottomOutlined />
            下载
          </IButton>
          <IButton type="dashed">
            <PlusOutlined />
            添加文件
          </IButton>
          <IButton>
            <UploadOutlined />
            导出
          </IButton>
          <IButton className="">增加时间</IButton>
          <IButton type="link">
            <EditOutlined />
            编辑
          </IButton>
          <IButton type="link">
            <UploadOutlined />
            上传
          </IButton>
        </div>
      </div>
      <div className="minor-row">
        <span className="minor-disabled">禁用状态</span>
        <div className="minor-item">
          <IButton disabled>
            <VerticalAlignBottomOutlined />
            下载
          </IButton>
          <IButton type="dashed" disabled>
            <PlusOutlined />
            添加文件
          </IButton>
          <IButton disabled>
            <UploadOutlined />
            导出
          </IButton>
          <IButton disabled>增加时间</IButton>
          <IButton type="link" disabled>
            <EditOutlined />
            编辑
          </IButton>
          <IButton type="link" disabled>
            <UploadOutlined />
            上传
          </IButton>
        </div>
      </div>
    </>
  );
};
```

<!-- <API src='../../src/components/IButton'> -->

### IButton prop 说明

| 属性      | 说明       | 类型                                                     | 默认值    | 是否必传 | 版本 |
| --------- | ---------- | -------------------------------------------------------- | --------- | -------- | ---- |
| className | 按钮的类名 | string                                                   |           | 否       |      |
| children  | 子元素     | React.ReactNode                                          |           | 否       |      |
| size      | 按钮大小   | 'xl' \| 'default' \| 'lg' \| 'md' \| 'sm'\| 'xs'\| 'xss' | 'default' | 否       |      |

IButton 是 antd button 的再封装，继承其属性，其他 API 详见https://ant.design/components/button-cn/#API
