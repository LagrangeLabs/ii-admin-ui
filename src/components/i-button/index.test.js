import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import IButton from './index';

Enzyme.configure({ adapter: new Adapter() });

describe("Test button component on the self's props", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should have the props class', () => {
    const wrapper = mount(<IButton size="lg">添加按钮</IButton>);

    expect(wrapper.find('.ii-button').length).toBeTruthy(); // 判断有没有对应的class

    expect(wrapper.find({ size: 'lg' }).length).toBeTruthy(); // 判断有没有size

    expect(wrapper.render()).toMatchSnapshot();
    wrapper.unmount();
  });
});
