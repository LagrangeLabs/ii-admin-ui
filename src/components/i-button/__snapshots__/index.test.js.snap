// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test button component on the self's props should have the props class 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "ant-btn ii-button",
      "data-testid": "test-button",
      "style": "height: 32px; padding: 0px 15px;",
      "type": "button",
    },
    "children": Array [
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "data": "添加按钮",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "data-testid": undefined,
      "style": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "data-testid": undefined,
      "style": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
